src/Features/CSharp/Portable/ConvertLinq/CSharpConvertForEachToLinqQueryProvider.cs

```diff
@@ -0,0 +1,585 @@
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.ConvertLinq;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertLinq
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = nameof(CSharpConvertForEachToLinqQueryProvider)), Shared]
+    internal sealed class CSharpConvertForEachToLinqQueryProvider : AbstractConvertForEachToLinqQueryProvider<ForEachStatementSyntax, StatementSyntax>
+    {
+        protected override ForEachStatementSyntax FindNodeToRefactor(SyntaxNode root, TextSpan span)
+            => root.FindNode(span) as ForEachStatementSyntax;
+
+        private static readonly TypeSyntax VarNameIdentifier = SyntaxFactory.IdentifierName("var");
+
+        protected override string Title => CSharpFeaturesResources.Convert_to_query;
+
+        protected override IConverter CreateDefaultConverter(
+            ForEachStatementSyntax forEachStatement,
+            ImmutableArray<SyntaxNode> convertingNodes,
+            ImmutableArray<SyntaxToken> identifiers,
+            ImmutableArray<StatementSyntax> statements)
+            => new DefaultConverter(forEachStatement, convertingNodes, identifiers, statements);
+
+        // Do not try to refactor queries with comments or conditional compilation in them.
+        // We can consider supporting queries with comments in the future.
+        protected override bool Validate(StatementSyntax statement)
+         => !(statement.ContainsDirectives ||
+            statement.DescendantTrivia().Any(trivia => trivia.MatchesKind(
+             SyntaxKind.SingleLineCommentTrivia,
+             SyntaxKind.MultiLineCommentTrivia,
+             SyntaxKind.MultiLineDocumentationCommentTrivia)));
+
+        protected override (ImmutableArray<SyntaxNode> ConvertingNodes, ImmutableArray<SyntaxToken> Identifiers, ImmutableArray<StatementSyntax> Statements) CreateForEachInfo(ForEachStatementSyntax forEachStatement)
+        {
+            var identifiers = new List<SyntaxToken>();
+            identifiers.Add(forEachStatement.Identifier);
+            var convertingNodes = new List<SyntaxNode>();
+            var current = forEachStatement.Statement;
+            IEnumerable<StatementSyntax> statementsCannotBeConverted = null;
+
+            void ProcessLocalDeclarationStatement(LocalDeclarationStatementSyntax localDeclarationStatement)
+            {
+                foreach (var variable in localDeclarationStatement.Declaration.Variables)
+                {
+                    convertingNodes.Add(variable);
+                    identifiers.Add(variable.Identifier);
+                }
+            }
+
+            // Setting statementsCannotBeConverted to anything means that we stop processing.
```

Comment:
this whole loop is pretty subtle (especially about pushing deeper into statements and whatnot).  Can a human-readonable explanation be given to given an intuition about this?